### PR TITLE
librbd: make CreatePrimaryRequest remove any unlinked mirror snapshots

### PR DESCRIPTION
**Pull request by idryomov**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by idryomov** [comment]
Fixed `MirrorSnapshot` (it was underspecified) and added `MirrorSnapshotOrphan` test.

File: qa/workunits/rbd/rbd_mirror_snapshot.sh [hunk 1]
```diff
@@ -220,7 +220,32 @@ wait_for_status_in_pool_dir ${CLUSTER1} ${POOL} ${image} 'up+replaying'
...
+  wait_for_status_in_pool_dir ${CLUSTER1} ${POOL} ${image} 'up+unknown'
+  wait_for_status_in_pool_dir ${CLUSTER2} ${POOL} ${image} 'up+unknown'
+  promote_image ${CLUSTER2} ${POOL} ${image}
+  wait_for_image_replay_started ${CLUSTER1} ${POOL} ${image}
+  wait_for_replay_complete ${CLUSTER1} ${CLUSTER2} ${POOL} ${image}
+  wait_for_status_in_pool_dir ${CLUSTER2} ${POOL} ${image} 'up+stopped'
+  wait_for_status_in_pool_dir ${CLUSTER1} ${POOL} ${image} 'up+replaying'
+done
+# check that demote (or other mirror snapshots) don't pile up
+test "$(count_mirror_snaps ${CLUSTER1} ${POOL} ${image})" -le 3
```
Comment (ajarr) [inline qa/workunits/rbd/rbd_mirror_snapshot.sh:245]: Maybe useful to comment why we expect the number of mirror snapshots to be <= 3?

**Review by ajarr — APPROVED** [review]
Code looks good to me

File: qa/workunits/rbd/rbd_mirror_snapshot.sh (hunk 1 above)
Comment (idryomov) [reply qa/workunits/rbd/rbd_mirror_snapshot.sh:245]: It's a bit hairy and based just on what rbd-mirror daemon happens to be doing today (not related to the changes made in https://github.com/ceph/ceph/pull/51166 and in this PR).  If you try some demotions and promotions, you would see that it doesn't unlink from/clean up everything that could be cleaned up.  The "cap" happens to be 3 snapshots: in the case of demoted image for example, it's the pre-demotion snapshot, demotion snapshot and non-primary snapshot synced following the other image getting promoted.

**Comment by idryomov** [comment]
No related failures (with Ubuntu 22.04 excluded due to a lab issue):

https://pulpito.ceph.com/dis-2023-09-03_10:29:06-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2023-09-04_00:20:56-rbd-wip-dis-testing-distro-default-smithi/

**Pull request merged**